### PR TITLE
commands: fix git version detection on macOS

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
I ran into this testing 0.11.0a1. Our current code matches against the
last word in the 'git --version' output. Unfortunately, for 'git
version 2.24.3 (Apple Git-128)', that's incorrectly matching against
'Git-128)', which doesn't work.

Generalize the version detection a bit to handle this.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
This is blocking v0.11. I'll backport once it's merged.

**Comment by marc-hb** [comment]
Just for the record:
```
$ /usr/bin/git --version
git version 2.24.3 (Apple Git-128)

$ /usr/local/bin/git --version # homebrew
git version 2.28.0
```

**Comment by marc-hb** [comment]
This sort of typical parsing issues is why I tend to prefer "trial and error" approaches like this one:
https://github.com/zephyrproject-rtos/west/compare/3c233fd14a7..339fe1a9ffa18e4a596#diff-8ab679fb4cff42c12eb05fbefdcebcebef2004540bf2b42bcb71ed5cede86a93L1083
```python
            try:
                # Silence the verbose and repetitive init.defaultBranch
                # warning. The branch 'placeholder' will never have any
                # commit so it will never actually exist.
                project.git(['init', '--initial-branch=placeholder',
                             project.abspath],
                            capture_stdout=True, capture_stderr=True,
                            cwd=util.west_topdir())
            except subprocess.CalledProcessError:
                # If git < 2.28 then the above failed because of
                # --initial-branch. If git >= 2.28 then it failed for
                # some other reason and will fail again for the same
                # reason which will be printed this time:
                project.git(['init', project.abspath], cwd=util.west_topdir())
```

File: src/west/commands.py
Comment (marc-hb) [inline src/west/commands.py:264]: Looks like this `^` has no effect when using `match()`, only when using `search()` (ignoring MULTILINE).

File: src/west/commands.py
Comment (marc-hb) [inline src/west/commands.py:266]: Have you considered `\b`oundary as in `r'\b(\d+)\.(\d+)\.(\d+)'` instead of the `.split()`? It's not equivalent as it would match `(2.8.1)` for instance but it would make the code much shorter.

File: src/west/commands.py
Comment (marc-hb) [reply src/west/commands.py:264]: Should the last digit be optional (`?`) to be on the safe side?

**Review by marc-hb — COMMENTED** [review]
Have you considered `from packaging import version` ?

**Comment by mbolivar-nordic** [comment]
> This sort of typical parsing issues is why I tend to prefer "trial and error" approaches like this one (PR #495 v1)

Heard, but this never hit any users since it was caught in pre-release testing, so I think it's okay to do it this way too.

**Comment by mbolivar-nordic** [comment]
> Have you considered `from packaging import version` ?

I hadn't, but we shouldn't be using that. The version.parse method is for parsing PEP 440 style versions, which are not guaranteed here.

File: src/west/commands.py
Comment (mbolivar-nordic) [reply src/west/commands.py]: > Looks like this `^` has no effect when using `match()`, only when using `search()` (ignoring MULTILINE).
> Should the last digit be optional (?) to be on the safe side?

Good catch on both fronts. I did some more digging in the git sources and pushed a new version that IMO handles all the possibilities we can reasonably expect.

File: src/west/commands.py
Comment (mbolivar-nordic) [reply src/west/commands.py]: > Have you considered `\b`oundary as in `r'\b(\d+)\.(\d+)\.(\d+)'` instead of the `.split()`? It's not equivalent as it would match `(2.8.1)` for instance but it would make the code much shorter.

I've golfed it down a few lines.

File: src/west/commands.py
Comment (marc-hb) [inline src/west/commands.py:302]: I guess you made the `\s` is non-optional based on what you learned in the sources? More precisely, you found it was easy to customize the version to `git version 6.6.6-whatever` but not to get rid of the `git version ` prefix?

File: src/west/commands.py
Comment (marc-hb) [inline src/west/commands.py:310]: Tiny nit, useless negation:
```python
if patch is None:
   return int(major), int(minor)
```

Alternatively:
```python
version = (major, minor)
if patch is not None
    version += (patch,)

return version
```

File: src/west/commands.py
Comment (marc-hb) [inline src/west/commands.py:305]: Nice.

File: src/west/commands.py
Comment (marc-hb) [inline src/west/commands.py:261]: I think this comment is useful but very long. The example and link are enough.

File: src/west/commands.py
Comment (marc-hb) [inline src/west/commands.py:273]: Just curious about the extra method. It's not that much code, even less without the comments.

**Review by marc-hb — APPROVED** [review]
As you spent time looking at the git sources I trust you didn't find any easier `git --parsable-version` ? Or at least none available in... old git versions :-D

PS: regular expressions are a programming language on their own but Python (and Perl?) bring that to a different level.

File: src/west/commands.py
Comment (mbolivar-nordic) [reply src/west/commands.py:273]: > Is the extra method to ease testing? Just curious.

Yes, exactly

File: src/west/commands.py
Comment (mbolivar-nordic) [reply src/west/commands.py:302]: > I guess you made the `\s` is non-optional based on what you learned in the git sources?

You made me triple check, and yes, not only is this safe, but it's safe regardless of the l10n (not i18n) settings: https://github.com/git/git/blob/7e391989789db82983665667013a46eabc6fc570/help.c#L646

I thought this could vary by locale, but I was misreading the gettext configuration files.

**Comment by mbolivar-nordic** [comment]
@marc-hb thanks for the approval. I've cleaned this up a bit per your review and will merge when CI passes.

**Comment by mbolivar-nordic** [comment]
> As you spent time looking at the git sources I trust you didn't find any easier `git --parsable-version` ? Or at least none available in... old git versions :-D

I addressed this in a response to a per-line comment, but I'll repeat it here. No, I can't find anything like that, but the help.c file makes it clear that we can expect git not to play tricks on us in the future by changing the way this works.

**Pull request merged**
